pkg/policycache/cache.go

```diff
@@ -89,131 +91,117 @@ func (m *pMap) add(policy *kyverno.ClusterPolicy) {
+				if enforcePolicy {
+					if !validateEnforceMap[kind+"/"+pName] {
+						validateEnforceMap[kind+"/"+pName] = true
+						if isNamespacedPolicy {
+							validatePolicy := m.kindDataMap[kind][ValidateEnforce]
+							m.kindDataMap[kind][ValidateEnforce] = append(validatePolicy, pName)
+							continue
+						}
+						validatePolicy := m.kindDataMap[kind][ValidateEnforce]
+						m.kindDataMap[kind][ValidateEnforce] = append(validatePolicy, policy.GetName())
+					}
 					continue
 				}
-				validatePolicy := m.dataMap[ValidateAudit]
-				m.dataMap[ValidateAudit] = append(validatePolicy, policy)
+
+				// ValidateAudit
+				if !validateAuditMap[kind+"/"+pName] {
+					validateAuditMap[kind+"/"+pName] = true
+					if isNamespacedPolicy {
+						validatePolicy := m.kindDataMap[kind][ValidateAudit]
+						m.kindDataMap[kind][ValidateAudit] = append(validatePolicy, pName)
+						continue
+					}
+					validatePolicy := m.kindDataMap[kind][ValidateAudit]
+					m.kindDataMap[kind][ValidateAudit] = append(validatePolicy, policy.GetName())
+				}
+				continue
 			}
-			continue
-		}
 
-		if rule.HasGenerate() {
-			if !generateMap[pName] {
-				generateMap[pName] = true
-				if isNamespacedPolicy {
-					generatePolicy := m.nsDataMap[policy.GetNamespace()][Generate]
-					m.nsDataMap[policy.GetNamespace()][Generate] = append(generatePolicy, policy)
-					continue
+			if rule.HasGenerate() {
+				if !generateMap[kind+"/"+pName] {
+					generateMap[kind+"/"+pName] = true
+					if isNamespacedPolicy {
+						generatePolicy := m.kindDataMap[kind][Generate]
+						m.kindDataMap[kind][Generate] = append(generatePolicy, pName)
+						continue
+					}
+					generatePolicy := m.kindDataMap[kind][Generate]
+					m.kindDataMap[kind][Generate] = append(generatePolicy, policy.GetName())
 				}
-				generatePolicy := m.dataMap[Generate]
-				m.dataMap[Generate] = append(generatePolicy, policy)
+				continue
 			}
-			continue
 		}
 	}
-
 	m.nameCacheMap[Mutate] = mutateMap
 	m.nameCacheMap[ValidateEnforce] = validateEnforceMap
 	m.nameCacheMap[ValidateAudit] = validateAuditMap
 	m.nameCacheMap[Generate] = generateMap
 }
 
-func (m *pMap) get(key PolicyType, nspace *string) []*kyverno.ClusterPolicy {
+func (m *pMap) get(key PolicyType, kind *string, nspace *string) (pname []string, allPolicies []*kyverno.ClusterPolicy) {
 	m.RLock()
 	defer m.RUnlock()
-	if nspace == nil || *nspace == "" {
-		return m.dataMap[key]
-	}
-	return m.nsDataMap[*nspace][key]
+	for _, policyName := range m.kindDataMap[*kind][key] {
+		ns, key, isNamespacedPolicy := policy2.ParseNamespacedPolicy(policyName)
+		if !isNamespacedPolicy {
+			policy, _ := m.pLister.Get(key)
+			allPolicies = append(allPolicies, policy)
+			pname = append(pname, policyName)
+		} else {
+			if ns == *nspace {
+				nspolicy, _ := m.npLister.Policies(ns).Get(key)
+				policy := policy2.ConvertPolicyToClusterPolicy(nspolicy)
+				allPolicies = append(allPolicies, policy)
+				pname = append(pname, key)
+			}
```

Comment:
As suggested above, let's not embed lister in `pMap`. Instead, let's create a separate method to get policy objects using lister.

So this method would look like:
`func (m *pMap) get(key PolicyType, kind string, nspace string) (pname []string)`

and the new one looks like this:
`func (pc *policyCache) getPolicyObject(kind, nspace string) []*kyverno.ClusterPolicy`